File: backend/services/notification_service.py
from typing import Dict, Any
from app.models import Order, User
from app.config import settings
import aiohttp
from datetime import datetime

class NotificationService:
    def __init__(self):
        self.email_api_key = settings.EMAIL_API_KEY
        self.sms_api_key = settings.SMS_API_KEY
        self.warehouse_email = settings.WAREHOUSE_EMAIL

    async def send_order_confirmation(self, order: Order) -> Dict[str, Any]:
        """Send order confirmation to customer"""
        customer = await User.get(order.user_id)
        if not customer:
            raise ValueError("Customer not found")

        # Prepare email content
        email_data = {
            "to": customer.email,
            "subject": f"Order Confirmation - #{order.id}",
            "template": "order_confirmation",
            "data": {
                "order_id": order.id,
                "customer_name": customer.name,
                "order_date": order.created_at.strftime("%Y-%m-%d %H:%M:%S"),
                "items": [
                    {
                        "name": item.name,
                        "quantity": item.quantity,
                        "price": item.price
                    }
                    for item in order.items
                ],
                "total": order.total,
                "shipping_address": order.shipping_address,
                "tracking_number": order.tracking_number
            }
        }

        # Send email
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{settings.EMAIL_API_URL}/send",
                headers={"Authorization": f"Bearer {self.email_api_key}"},
                json=email_data
            ) as response:
                if response.status != 200:
                    raise ValueError("Failed to send order confirmation email")

        # Send SMS if phone number is available
        if customer.phone:
            sms_data = {
                "to": customer.phone,
                "message": f"Your order #{order.id} has been confirmed. Tracking number: {order.tracking_number}"
            }

            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f"{settings.SMS_API_URL}/send",
                    headers={"Authorization": f"Bearer {self.sms_api_key}"},
                    json=sms_data
                ) as response:
                    if response.status != 200:
                        print("Failed to send SMS notification")

        return {"status": "success", "order_id": order.id}

    async def notify_warehouse(self, order: Order) -> Dict[str, Any]:
        """Notify warehouse about new order"""
        email_data = {
            "to": self.warehouse_email,
            "subject": f"New Order - #{order.id}",
            "template": "warehouse_notification",
            "data": {
                "order_id": order.id,
                "items": [
                    {
                        "name": item.name,
                        "quantity": item.quantity,
                        "sku": item.sku
                    }
                    for item in order.items
                ],
                "shipping_address": order.shipping_address,
                "priority": order.priority
            }
        }

        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{settings.EMAIL_API_URL}/send",
                headers={"Authorization": f"Bearer {self.email_api_key}"},
                json=email_data
            ) as response:
                if response.status != 200:
                    raise ValueError("Failed to notify warehouse")

        return {"status": "success", "order_id": order.id}

    async def send_status_update(self, order: Order) -> Dict[str, Any]:
        """Send order status update to customer"""
        customer = await User.get(order.user_id)
        if not customer:
            raise ValueError("Customer not found")

        email_data = {
            "to": customer.email,
            "subject": f"Order Status Update - #{order.id}",
            "template": "status_update",
            "data": {
                "order_id": order.id,
                "customer_name": customer.name,
                "new_status": order.status,
                "tracking_number": order.tracking_number,
                "update_time": datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
            }
        }

        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{settings.EMAIL_API_URL}/send",
                headers={"Authorization": f"Bearer {self.email_api_key}"},
                json=email_data
            ) as response:
                if response.status != 200:
                    raise ValueError("Failed to send status update email")

        return {"status": "success", "order_id": order.id}

    async def send_cancellation_notice(self, order: Order) -> Dict[str, Any]:
        """Send order cancellation notice to customer"""
        customer = await User.get(order.user_id)
        if not customer:
            raise ValueError("Customer not found")

        email_data = {
            "to": customer.email,
            "subject": f"Order Cancelled - #{order.id}",
            "template": "cancellation_notice",
            "data": {
                "order_id": order.id,
                "customer_name": customer.name,
                "cancellation_time": datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S"),
                "refund_amount": order.total
            }
        }

        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{settings.EMAIL_API_URL}/send",
                headers={"Authorization": f"Bearer {self.email_api_key}"},
                json=email_data
            ) as response:
                if response.status != 200:
                    raise ValueError("Failed to send cancellation notice")

        return {"status": "success", "order_id": order.id}

    async def send_low_stock_alert(self, product_id: str) -> Dict[str, Any]:
        """Send low stock alert to warehouse"""
        product = await Product.get(product_id)
        if not product:
            raise ValueError("Product not found")

        email_data = {
            "to": self.warehouse_email,
            "subject": f"Low Stock Alert - {product.name}",
            "template": "low_stock_alert",
            "data": {
                "product_id": product.id,
                "product_name": product.name,
                "current_stock": product.stock,
                "minimum_stock": 10,  # Assuming 10 is the minimum stock threshold
                "alert_time": datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
            }
        }

        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{settings.EMAIL_API_URL}/send",
                headers={"Authorization": f"Bearer {self.email_api_key}"},
                json=email_data
            ) as response:
                if response.status != 200:
                    raise ValueError("Failed to send low stock alert")

        return {"status": "success", "product_id": product_id} 

File: frontend/src/services/notificationService.js
import api from './api';

const notificationService = {
    /**
     * Get notifications with optional filtering
     * @param {Object} params - Filter parameters
     * @param {string} [params.type] - Filter by notification type
     * @param {string} [params.status] - Filter by notification status
     * @param {Date} [params.startDate] - Filter by start date
     * @param {Date} [params.endDate] - Filter by end date
     * @param {number} [params.limit=100] - Number of notifications to return
     * @param {number} [params.skip=0] - Number of notifications to skip
     * @returns {Promise<Array>} List of notifications
     */
    getNotifications: async (params = {}) => {
        const response = await api.get('/notifications', { params });
        return response.data;
    },

    /**
     * Get a specific notification by ID
     * @param {string} notificationId - The ID of the notification
     * @returns {Promise<Object>} The notification
     */
    getNotification: async (notificationId) => {
        const response = await api.get(`/notifications/${notificationId}`);
        return response.data;
    },

    /**
     * Create a new notification
     * @param {Object} notification - The notification data
     * @returns {Promise<Object>} The created notification
     */
    createNotification: async (notification) => {
        const response = await api.post('/notifications', notification);
        return response.data;
    },

    /**
     * Update notification status
     * @param {string} notificationId - The ID of the notification
     * @param {string} status - The new status
     * @param {string} [error] - Error message if any
     * @returns {Promise<Object>} The updated notification
     */
    updateNotificationStatus: async (notificationId, status, error) => {
        const response = await api.put(`/notifications/${notificationId}/status`, {
            status,
            error
        });
        return response.data;
    },

    /**
     * Delete a notification
     * @param {string} notificationId - The ID of the notification
     * @returns {Promise<Object>} Success status
     */
    deleteNotification: async (notificationId) => {
        const response = await api.delete(`/notifications/${notificationId}`);
        return response.data;
    },

    /**
     * Get notification statistics
     * @returns {Promise<Object>} Notification statistics
     */
    getNotificationStats: async () => {
        const response = await api.get('/notifications/stats/summary');
        return response.data;
    },

    /**
     * Send order confirmation notification
     * @param {string} orderId - The ID of the order
     * @returns {Promise<Object>} The result of the notification
     */
    sendOrderConfirmation: async (orderId) => {
        const response = await api.post(`/notifications/orders/${orderId}/confirm`);
        return response.data;
    },

    /**
     * Notify warehouse about new order
     * @param {string} orderId - The ID of the order
     * @returns {Promise<Object>} The result of the notification
     */
    notifyWarehouse: async (orderId) => {
        const response = await api.post(`/notifications/orders/${orderId}/notify-warehouse`);
        return response.data;
    },

    /**
     * Send order status update notification
     * @param {string} orderId - The ID of the order
     * @returns {Promise<Object>} The result of the notification
     */
    sendStatusUpdate: async (orderId) => {
        const response = await api.post(`/notifications/orders/${orderId}/status-update`);
        return response.data;
    },

    /**
     * Send order cancellation notice
     * @param {string} orderId - The ID of the order
     * @returns {Promise<Object>} The result of the notification
     */
    sendCancellationNotice: async (orderId) => {
        const response = await api.post(`/notifications/orders/${orderId}/cancel`);
        return response.data;
    },

    /**
     * Send low stock alert
     * @param {string} productId - The ID of the product
     * @returns {Promise<Object>} The result of the notification
     */
    sendLowStockAlert: async (productId) => {
        const response = await api.post(`/notifications/products/${productId}/low-stock`);
        return response.data;
    }
};

export default notificationService; 